content/blog/2020-12-22-pipeline-II-analysis/index.md

```diff
@@ -0,0 +1,150 @@
+
++++
+title = """Loop Pipeline Initiation Interval Estimation Using LLVM"""
+[extra]
+bio = """
+[Jie Liu](https://github.com/Sibylau) is a 2nd year PhD student in Computer Systems Lab at Cornell. She is working on domain-specific language and accelerators.
+"""
+latex = true
+[[extra.authors]]
+name = "Jie Liu"
+link = "https://github.com/Sibylau"
++++
+
+## Introdution
+Loop pipelining is an important performance optimization technique that exploits 
+the parallelism among loop iterations. In unoptimized loops written in sequential languages 
+like C/C++, one iteration can only begin after the previous iteration is complete. 
+Loop pipelining allows loop iterations to overlap, which better utilizes resources and 
+improves the overall throughput. To implement loop pipelining in hardware, however, it 
+usually takes non-trivial efforts for hardware-description languages (HDL) programmers 
+to produce a cycle-accurate register-transfer-level (RTL) design. Moreover, once the RTL 
+design is fixed, it is difficult to retarget various design points because the datapath and 
+control logics are explicitly constructed in the implementation.
+
+With high-level synthesis (HLS) emerging as an alternative design fashion, programmers 
+can simply add a pragma in a C/C++ loop to enable loop pipelining in hardware. However,
+programmers at an early stage don't have much information on the performance of loops 
+they write. If the specified pipeline initial interval (II) cannot be achieved, the HLS 
+tools will throw out messages at least after running synthesis, which can take hours for 
+a large design. In this project, I want to perform a quick analysis of the achievable loop 
+pipeline II written in C-level. I believe early feedbacks will help programmers iterate 
+their designs in a much faster way.
+
+## Background
+There are typically two factors of constraints that limit the degree of parallelism loop 
+pipelining can exploit: one is hardware resource contention, and the other is the data 
+dependencies between loop iterations. 
+
+We denote the lower-bound II limited by resource usage confliction as ResMII, which is 
+given by the formula (1):
+<img src="https://github.com/Sibylau/cs6120/blob/project-jl3952/content/blog/2020-12-22-pipeline-II-analysis/mii.PNG" width="300" />

```

Comment:
Using an image is fine, but FWIW, you could also put this right here as embedded LaTeX math.